client/state/connectionMiddleware/middleware.js

```diff
@@ -0,0 +1,120 @@
+/**
+ * External dependencies
+ */
+import uniqueId from 'lodash/utility/uniqueId';
+
+/**
+ * Internal dependencies
+ */
+import {
+	isOffline,
+	isOnline
+} from 'state/application/selectors';
+import {
+	OFFLINE_QUEUE_ACTION,
+	CONNECTION_RESTORED,
+	OFFLINE_QUEUE_REMOVE,
+} from 'state/action-types';
+const ACTION_TIMEOUT = 10000;
+
+const connections = {};
+
+let actionTimeout = null;
+
+//Prob should be in actions
+function queueAction( action ) {
+	return {
+		type: OFFLINE_QUEUE_ACTION,
+		id: uniqueId(),
+		dispatchedAt: ( new Date() ).getTime(),
+		action: action
```

Comment:
@mtias could you give an example of why we should rearrange action order? It would be a bit difficult to guarantee correctness for modifying actions.